File: apps/goatos-android/feature/feature-calendar/src/main/kotlin/sg/mesha/goatos/feature/calendar/DriveCardMetrics.kt
package sg.mesha.goatos.feature.calendar

// Pure, unit-tested metrics for the park-level vaccination drive card. Extracted from
// CalendarScreen.kt so the coverage-percentage rounding and the drive-summary visibility
// rules are testable on the JVM (DriveCardMetricsTest) and cannot silently regress (CDR-005).

// Coverage-ring percentage on the DISTINCT-ANIMAL grain, rounded half-up to match the web
// card (which uses Math.round). Example: 46/77 -> 60, NOT the truncated 59 the pre-fix
// integer-division card produced. Returns 0 when there are no animals.
internal fun driveCoveragePct(completedAnimals: Int, totalAnimals: Int): Int =
    if (totalAnimals > 0) Math.round(completedAnimals * 100f / totalAnimals) else 0

// A drive event's flat subtitle/summary rows render ONLY when there is no backend
// drive_summary. When a summary is present the DriveProgressCard renders the sheds/
// vaccines/dose totals, so the flat rows would duplicate them (CDR-003). Returns true when
// the legacy rows should show (summary absent) and false when they must be hidden (present).
internal fun legacyDriveRowsVisible(drive: CalendarDriveSummary?): Boolean = drive == null

// The coverage ring's numerator/denominator + which grain it represents.
internal data class DriveCoverage(val completed: Int, val total: Int, val usesAnimals: Boolean)

// A status chip for the redesigned card: one bucket (done/due/overdue/deferred) with a count.
internal data class StatusChip(val key: String, val count: Int, val label: String)

// Distinct-animal counts are authoritative for the ring, BUT a Room cache written before the
// total_animals/completed_animals fields shipped -- or a mixed-version API response mid-rollout --
// has NO animal counts. They decode as null (not 0). Rendering them as "0 / 0 animals" over a drive
// with valid legacy dose counts is a false empty state (CDR-R1). When either animal count is absent,
// fall back to the obligation (dose) counts that ARE present, honestly labelled; a background refresh
// then repopulates the animal grain.
internal fun driveCoverage(
    completedAnimals: Int?,
    totalAnimals: Int?,
    completedDoses: Int,
    totalDoses: Int,
): DriveCoverage =
    if (completedAnimals != null && totalAnimals != null) {
        DriveCoverage(completedAnimals, totalAnimals, usesAnimals = true)
    } else {
        DriveCoverage(completedDoses, totalDoses, usesAnimals = false)
    }

// The drive progress the card renders. The backend owns the numerator, the denominator, and the
// GRAIN they are counted on (progress_basis), and both clients render it verbatim -- this is the
// cross-surface parity contract. Before it existed, this function returned max(submitted, completed)
// while the admin-web card used completed only, so the SAME drive showed two different completion
// numbers and two different ring percentages. Submitted-but-unverified work is NOT progress; it
// stays visible through submittedAnimals/submittedCount and the "verification pending" strip.
//
// The local derivation below is a LEGACY fallback only, for a Room row cached before these fields
// shipped or a mixed-version response from an older backend (they decode as null, not 0). A
// background refresh replaces it with the contract values.
internal fun driveVisibleProgress(summary: CalendarDriveSummary): DriveCoverage {
    val completed = summary.progressCompleted
    val total = summary.progressTotal
    if (completed != null && total != null) {
        return DriveCoverage(completed, total, usesAnimals = summary.progressBasis != "doses")
    }
    return driveCoverage(
        completedAnimals = summary.completedAnimals,
        totalAnimals = summary.totalAnimals,
        completedDoses = summary.completedCount,
        totalDoses = summary.totalCount,
    )
}

// The percentage shown in the ring. Backend-rounded value wins verbatim so the ring reads
// identically on web and mobile; the local half-up rounding is the same legacy fallback.
internal fun drivePctFor(summary: CalendarDriveSummary, coverage: DriveCoverage): Int =
    summary.progressPct ?: driveCoveragePct(coverage.completed, coverage.total)

// Status chips for the drive card: the nonzero buckets in fixed order, each with a key, count and
// placeholder label. Labels are localized at render time via stringResource(). Returns empty list
// if all buckets are zero (empty card state).
//
// Maintainer decision 2026-08-03: NO "submitted" chip. The card already carries a
// "Verification pending" status chip for exactly that state, so a second "20 submitted" chip says
// the same thing one line lower. The submitted count is not lost -- it is the progress numerator,
// so a fully-submitted drive reads 100% with the status chip naming what is still outstanding.
// Do not re-add it "for completeness".
internal fun driveStatusChips(summary: CalendarDriveSummary): List<StatusChip> =
    listOfNotNull(
        if (summary.completedCount > 0) StatusChip("completed", summary.completedCount, "Completed") else null,
        if (summary.dueCount > 0) StatusChip("due", summary.dueCount, "Due") else null,
        if (summary.overdueCount > 0) StatusChip("overdue", summary.overdueCount, "Overdue") else null,
        if (summary.deferredCount > 0) StatusChip("deferred", summary.deferredCount, "Deferred") else null,
    )

File: apps/admin-web/features/calendar/drive-card-metrics.ts
// Pure, unit-tested metrics for the park-level vaccination drive card. Extracted from
// calendar-drive-card.tsx so the coverage-percentage rounding cannot silently regress
// (CDR-005). Covered by drive-card-metrics.test.mjs.

// Coverage-ring percentage on the DISTINCT-ANIMAL grain, rounded half-up to match the
// Android card (CalendarScreen.kt also uses Math.round). Example: 46/77 -> 60 (NOT the
// truncated 59 the pre-fix Android card produced). Returns 0 when there are no animals.
export function driveCoveragePct(completedAnimals: number, totalAnimals: number): number {
  if (totalAnimals <= 0) return 0;
  return Math.round((completedAnimals / totalAnimals) * 100);
}

export function driveClosedCoveragePct(completedAnimals: number, totalAnimals: number, eventStatus: string | null | undefined): number {
  if (eventStatus === "completed") return driveCoveragePct(completedAnimals, totalAnimals);
  if (totalAnimals <= 0 || completedAnimals <= 0) return 0;
  return Math.min(99, driveCoveragePct(completedAnimals, totalAnimals));
}

export interface DriveCoverage {
  completed: number;
  total: number;
  usesAnimals: boolean;
}

// Distinct-animal counts are authoritative for the ring, BUT a mixed-version API response mid-rollout
// (an older backend instance that predates the total_animals/completed_animals fields) can omit them --
// they arrive as undefined, NOT 0. Rendering "0 / 0 animals" over a drive with valid dose counts is a
// false empty state (CDR-R1). When either animal count is missing, fall back to the obligation (dose)
// counts that ARE present, labelled accordingly; a later response repopulates the animal grain.
export function driveCoverage(
  completedAnimals: number | null | undefined,
  totalAnimals: number | null | undefined,
  completedDoses: number,
  totalDoses: number,
): DriveCoverage {
  if (typeof completedAnimals === "number" && typeof totalAnimals === "number") {
    return { completed: completedAnimals, total: totalAnimals, usesAnimals: true };
  }
  return { completed: completedDoses, total: totalDoses, usesAnimals: false };
}

// The drive progress the card renders. The backend owns the numerator, the denominator and the
// GRAIN they are counted on (progress_basis), and BOTH clients render it verbatim -- this is the
// cross-surface parity contract. Before it existed this card used completed_animals while the
// Android card used max(submitted, completed), so the SAME drive showed two different completion
// numbers and two different ring percentages. The numerator is verified completion only; submitted-
// but-unverified work stays visible through submitted_animals and the verification-pending strip.
// The local derivation is a LEGACY fallback for a mixed-version response from an older backend that
// predates these fields (they arrive undefined, not 0).
export function driveVisibleProgress(summary: {
  progress_basis?: string | null;
  progress_completed?: number | null;
  progress_total?: number | null;
  completed_animals?: number | null;
  total_animals?: number | null;
  completed_count: number;
  total_count: number;
}): DriveCoverage {
  if (typeof summary.progress_completed === "number" && typeof summary.progress_total === "number") {
    return {
      completed: summary.progress_completed,
      total: summary.progress_total,
      usesAnimals: summary.progress_basis !== "doses",
    };
  }
  return driveCoverage(summary.completed_animals, summary.total_animals, summary.completed_count, summary.total_count);
}

// Backend-rounded percentage wins verbatim so the ring reads identically on web and mobile; the
// local half-up rounding is the same legacy fallback.
export function drivePctFor(summary: { progress_pct?: number | null }, coverage: DriveCoverage): number {
  if (typeof summary.progress_pct === "number") return summary.progress_pct;
  return driveCoveragePct(coverage.completed, coverage.total);
}

export interface DriveStatusChip {
  key: "completed" | "submitted" | "due" | "overdue" | "deferred";
  count: number;
}

export function driveStatusClass(key: DriveStatusChip["key"]): "completed" | "due" | "over" | "def" {
  if (key === "overdue") return "over";
  if (key === "deferred") return "def";
  if (key === "submitted") return "due";
  return key;
}

// Returns all nonzero status buckets in fixed order: completed, due, overdue, deferred.
// Used to render the status chips row on the drive card.
export function driveStatusChips(summary: {
  completed_count: number;
  submitted_count?: number;
  due_count: number;
  overdue_count: number;
  deferred_count: number;
}): DriveStatusChip[] {
  return (
    [
      { key: "completed" as const, count: summary.completed_count },
      { key: "submitted" as const, count: summary.submitted_count ?? 0 },
      { key: "due" as const, count: summary.due_count },
      { key: "overdue" as const, count: summary.overdue_count },
      { key: "deferred" as const, count: summary.deferred_count },
    ] as const
  ).filter((chip) => chip.count > 0);
}
